core/service/src/lib.rs

```diff
@@ -746,19 +761,27 @@ fn build_network_future<
 		}
 
 		// Interval report for the external API.
-		while let Ok(Async::Ready(_)) = status_interval.poll() {
-			let status = NetworkStatus {
-				sync_state: network.sync_state(),
-				best_seen_block: network.best_seen_block(),
-				num_sync_peers: network.num_sync_peers(),
-				num_connected_peers: network.num_connected_peers(),
-				num_active_peers: network.num_active_peers(),
-				average_download_per_sec: network.average_download_per_sec(),
-				average_upload_per_sec: network.average_upload_per_sec(),
-			};
-			let state = network.network_state();
-
-			status_sinks.lock().retain(|sink| sink.unbounded_send((status.clone(), state.clone())).is_ok());
+		{
+			let mut status_sinks = status_sinks.lock();
+			for n in (0..status_sinks.len()).rev() {
+				while let Ok(Async::Ready(_)) = status_sinks[n].0.poll() {
```

Comment:
This should use `FuturesUnordered`.

Reply:
I'm not sure how you see that. The intervals are streams and not futures, and I would have to maintain a parallel list of senders.

Reply:
Yeah, just implement your own future. 
Above you have written that this outer future is called "very very often" and every time you will iterate the list of status sinks and ask "hey is your timer finished?".
`FuturesUnordered` is exactly for such use cases as it will not iterate the full list every time we poll this future.

```
struct StatusSink {
    sender: Option<Sender>,
    timer: Option<Timer>,
}

impl StatusSink {
    fn send_status(&mut self, status) -> bool {
        self.sender.as_mut().unwrap().unbounded_send(status).is_err()
    }
}

impl Future for StatusSink {
    type Item = Self;
    type Error = ();

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        match self.timer.as_mut().unwrap().poll()? {
            Async::Ready(_) => StatusSink { sender: self.sender.take(), timer: self.timer.take() },
            Async::NotReady => NotReady,
        }
    }
}
```